### ISSUE TITLE
Analyzer does not propagate non-nullability from `?..` in cascade sequence

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#39209]
In a cascade sequence of the form `a?..<cascade1>..<cascade2>`, the analyzer currently issues an error on the second cascade entry.  This is incorrect, the initial null check on `a` applies to the entire cascade sequence.

```dart
void main() {
  int? x = null;
  x?..isEven; // ok
  x?..isEven..isEven; // error, should be ok.
}
```

This is perhaps under-specified in the feature spec, I've filed [an issue](https://github.com/dart-lang/language/issues/659) to make this clear.

cc @scheglov @stereotype441

**Comment by stereotype441** [comment]
I just spent some time digging into the flow analysis implications of this.  The resolver needs to call FlowAnalysis.nullAwareAccess_begin each time an operator like `?..` is encountered, and FlowAnalysis.nullAwareAccess_end whenever the null-shorting influence of the `?..` ends.  So if I'm understanding things correctly (see my clarifying question [here](https://github.com/dart-lang/language/issues/659#issuecomment-550254729)), in the `x?..isEven..isEven` example, the sequence of operations should be:
- Visit `x`
- Call `FlowAnalysis.nullAwareAccess_begin`
- Visit the first `isEven`
- Visit the second `isEven`
- Call `FlowAnalysis.nullAwareAccess_end`

And in an example like `x?..isEven?..isEven`, the sequence of operations should be:
- Visit `x`
- Call `FlowAnalysis.nullAwareAccess_begin`
- Visit the first `isEven`
- Call `FlowAnalysis.nullAwareAccess_begin` (*)
- Visit the second `isEven`
- Call `FlowAnalysis.nullAwareAccess_end` (*)
- Call `FlowAnalysis.nullAwareAccess_end`

(however, since the second `?` is redundant, the calls marked with (*) could be dropped without any visible effect).

A user-visible consequence of this is that in the following code:
```
class C {
  dynamic x;
  void f(dynamic value) {}
}
void test(C? c) {
  c?..f(c.x)  // (1)
    ..f(c.x); // (2)
}
```
No null checks are needed for the subexpressions `c.x` at `(1)` and `(2)`, because the `?..` causes `c` to be promoted to a non-nullable `C` for the duration of the whole cascade sequence.

**Comment by leafpetersen** [comment]
> No null checks are needed for the subexpressions `c.x` at `(1)` and `(2)`, because the `?..` causes `c` to be promoted to a non-nullable `C` for the duration of the whole cascade sequence.

Agreed.  Each cascade section is valid, and within each cascade section, the target of the cascade, if it is a promotion candidate, has been promoted.

**Comment by leafpetersen** [comment]
This also seems to hit some sub-parts of cascades sequences.  Oddly it only seems to affect method calls, and not getter calls.

```dart
class A {
  List<int> foo() => [];
  List<int> get bar => [];
}

main() {
  A? a = A();
  a?..foo().length;  // Incorrectly gives an error
  a?..bar.length;    // Correctly gives no error
}
```

**Comment by scheglov** [comment]
Fixed as 4844fa96223e2c7af2893608328df4f604fe510a.
